README.md
# Google Maps Marker Animation - Overview

<a href="https://www.buymeacoffee.com/gauris" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" width="195" height="55"></a>

<a href="https://pub.dev/packages/flutter_animarker">
    <img src="https://img.shields.io/badge/pub-3.2.0-blue" alt="pub package"/>
</a>

Sometimes you need more than place a *Marker*📍 at map canvas 🌍, you need to smoothly move through **Google Maps**.

This package will help you to animate *Markers*' position changes and more.

This version v3.0.0 includes many useful features, namely:

 - Marker's position animation
 - Multiple markers' animations at the same time
 - Null-safety compatible
 - Ripple effect over marker position
 - Marker's rotation or bearing/heading of direction
 - Multipoint linear animation (*Piecewise Linear Approximation Algorithm*)
 - Support animation curves and duration
 - Widget-based with fully customized behaviors
 - Animation warm-up for improving performance
 - Useful **LocationTween**, **AngleTween** and **PolynomialLocationInterpolator** core logic

### Note
This package only animate the marker's changes. Both Geolocation and Google Maps configuration are out of the scope of this package. So, before trying this package to ensure that you place a marker you Google Map or get location updates.


## Screenshots

<img src="https://raw.githubusercontent.com/gauris26/flutter_animarker/d61ac4f420f030f4e002fa287282628d901cff26/arts/marker_animation.gif" width="300"/> <img src="https://raw.githubusercontent.com/gauris26/flutter_animarker/d61ac4f420f030f4e002fa287282628d901cff26/arts/multi_markers.gif" width="300"/>

## Example

```dart
  Animarker(
    ...
    curve: Curves.bounceInOut,
    duration: Duration(milliseconds: 2000),
    markers: <Marker> {
      RippleMarker(
        markerId: MarkerId('MarkerId2'),
        position: LatLng(0, 0),
        ripple: true,
      ),
    },
    child: GoogleMap(
      ...
      onMapCreated: (gController) => controller.complete(gController), //Complete the future GoogleMapController
      ...
    ),
    ...
  )
```

```dart
    //Setting dummies values
    const kStartPosition = LatLng(18.488213, -69.959186);
    const kSantoDomingo = CameraPosition(target: kStartPosition, zoom: 15);
    const kMarkerId = MarkerId('MarkerId1');
    const kDuration = Duration(seconds: 2);
    const kLocations = [
      kStartPosition,
      LatLng(18.488101, -69.957995),
      LatLng(18.489210, -69.952459),
      LatLng(18.487307, -69.952759)
    ];

    class SimpleMarkerAnimationExample extends StatefulWidget {
      @override
      SimpleMarkerAnimationExampleState createState() => SimpleMarkerAnimationExampleState();
    }

    class SimpleMarkerAnimationExampleState extends State<SimpleMarkerAnimationExample> {
      final markers = <MarkerId, Marker>{};
      final controller = Completer<GoogleMapController>();
      final stream = Stream.periodic(kDuration, (count) => kLocations[count]).take(kLocations.length);

      @override
      void initState() {
        stream.forEach((value) => newLocationUpdate(value));

        super.initState();
      }

      @override
      Widget build(BuildContext context) {
        return MaterialApp(
          title: 'Google Maps Markers Animation Example',
          home: Animarker(
            curve: Curves.ease,
            mapId: controller.future.then<int>((value) => value.mapId), //Grab Google Map Id
            markers: markers.values.toSet(),
            child: GoogleMap(
              mapType: MapType.normal,
              initialCameraPosition: kSantoDomingo,
              onMapCreated: (gController) => controller.complete(gController), //Complete the future GoogleMapController
            ),
          ),
        );
      }

      void newLocationUpdate(LatLng latLng) {
         var marker = RippleMarker(
           markerId: kMarkerId,
           position: latLng,
           ripple: true,
         );
         setState(() => markers[kMarkerId] = marker);
      }
    }
 ```
## Using Ripple Effect

You only need to use the marker wrapper class ``` RippleMarker```, and set the ripple ```flag```  in order you can the ripple state.

```dart
  Animarker(
    // Other properties
    rippleRadius: 0.5,  //[0,1.0] range, how big is the circle
    rippleColor: Colors.teal, // Color of fade ripple circle
    rippleDuration: Duration(milliseconds: 2500), //Pulse ripple duration
    markers: <Marker>{
      //Ripple Marker
      RippleMarker(
        markerId: MarkerId('MarkerId1'),
        position: LatLng(0, 0),
        ripple: true,  //Ripple state
      ),
      //Non-ripple marker
      Marker(
        markerId: MarkerId('MarkerId2'),
        position: LatLng(0, 0),
      ),
    },
    // Other properties
  )
```
So, Let's it rip!

## Using Rotation

The Marker rotation can be useful for Uber-like or delivery apps. The bearing or heading is the angle of direction which the *Marker* is moving toward to over the earth.

```dart
  Animarker(
    // Other properties
    useRotation : true, // Actived by default
    markers: <Marker>{
      Marker(
      markerId: MarkerId('MarkerId2'),
      position: LatLng(0, 0),
      )
    },
  // Other properties
  )
```
This way ```useRotation = true```, you control globally if *Marker* should rotate or not.

## Using Curve and Duration

Just like a normal *Flutter* animation, you can set a ```Curve``` o ```Duration``` to get the desire effect or result. So flexible, right?

```dart
  Animarker(
    // Other properties
    curve: Curves.bounceInOut,
    duration: Duration(milliseconds: 2000),
    markers: <Marker>{
      Marker(
      markerId: MarkerId('MarkerId2'),
      position: LatLng(0, 0),
      ),
    },
    // Other properties
  )
```

## License

 BSD-3-Clause License   Copyright 2020  Gauris Javier. All rights
  reserved.
   Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions  are met:
   1. Redistributions of source code must retain the above copyright
  notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright
 notice, this list of conditions and the following disclaimer in the  documentation and/or other materials provided with the distribution.
   3. Neither the name of the copyright holder nor the names of its contributors
 may be used to endorse or promote products derived from this software  without specific prior written permission.
   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.```


CHANGELOG.md
# 4.0.0
    - This version upgrades all dependencies to match the latest Dart and Flutter versions, to start addressing bugs, and introduces new features then.
# 3.3.1-beta.4
    - runExpressAfter property to control when the animation queue should run express to the last position
    - shouldAnimateCamera property to prevent Google Map Camera Updates to be handled
    - rippleIdleAfter property to stop ripple animation on static marker (no updates) after timeout
    - Fixing bug when trying to clear the markers
    - Improve Ripple effect during Marker movement.

# 3.2.0
    - onTap() marker function support added
    - Added a flag to control when animate Google Maps Camera.
    - Add extended documentation for Animarker properties
    - Fix: Solve flicking issue when user location is moving fast
    - Fix: Autostart animation when the source of location changes stop.

# 3.2.0-beta.3
    - onTap() marker function support added

# 3.1.3-beta.2
    - Added a flag to control when animate Google Maps Camera.

# 3.1.2-beta.1
    - Add extended documentation for Animarker properties
    - Fix: Solve flicking issue when user location is moving fast
    - Fix: Autostart animation when the source of location changes stop.

# 3.1.1-alpha.2
    - Trying with HTML Markup to resize image

# 3.1.1-alpha.1
    - Minor: GIF image resizing

# 3.1.1+2-alpha

**BREAKING**: The library will use now a better reactive Flutter approach
 - Marker's position animation
 - Multiple markers' animations at the same time
 - Null-safety compatible
 - Ripple effect over marker position
 - Marker's rotation or bearing/heading of direction
 - Multipoint linear animation (*Piecewise Linear Approximation Algorithm*)
 - Support animation curves and duration
 - Widget-based with fully customized behaviors
 - Animation warm-up for improving performance
 - Useful **LocationTween**, **AngleTween** and **PolynomialLocationInterpolator** core logic
 - More test coverage

# 1.0.0

- Remove Google Maps dependencies
- Added multiple markers at same time functionality

# 0.1.2

- Update Google Maps constraints and pull request merging

# 0.1.1+1

- Update metadata plugin information

# 0.1.1

- Added working project example to the GitHub repo. Just add your own **Google Maps Api Key**.

# 0.0.2

- Update with a working example and GIF showing the animation movement

# 0.0.1+3

- Issue setting the Duration interval of interpolation

# 0.0.1+2

- Minor error corrections

# 0.0.1+1

- Interpolation between two position points, including rotation of the marker
- Added examples and better description

# 0.0.1

- Interpolation between two position points, including rotation of the marker



















example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_animarker/flutter_map_marker_animation.dart';
import 'package:flutter_config/flutter_config.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

import 'listen_locations_updates.dart';

//Setting dummies values
const kStartPosition = LatLng(18.488213, -69.959186);
const kSantoDomingo = CameraPosition(target: kStartPosition, zoom: 15);
const kMarkerId = MarkerId('MarkerId1');
const kDuration = Duration(seconds: 2);
const kLocations = [
  kStartPosition,
  LatLng(18.488101, -69.957995),
  LatLng(18.489210, -69.952459),
  LatLng(18.487307, -69.952759),
  LatLng(18.487308, -69.952759),
];

void main() async {
  WidgetsFlutterBinding.ensureInitialized(); // Required by FlutterConfig
  await FlutterConfig.loadEnvVariables();

  runApp(SimpleMarkerAnimationExample());
}

class SimpleMarkerAnimationExample extends StatefulWidget {
  @override
  SimpleMarkerAnimationExampleState createState() =>
      SimpleMarkerAnimationExampleState();
}

class SimpleMarkerAnimationExampleState
    extends State<SimpleMarkerAnimationExample> {
  final markers = <MarkerId, Marker>{};
  final controller = Completer<GoogleMapController>();
  final stream = Stream.periodic(kDuration, (count) => kLocations[count])
      .take(kLocations.length);

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Google Maps Markers Animation Example',
      home: Animarker(
        curve: Curves.bounceOut,
        rippleRadius: 0.2,
        useRotation: false,
        duration: Duration(milliseconds: 2300),
        mapId: controller.future
            .then<int>((value) => value.mapId), //Grab Google Map Id
        markers: markers.values.toSet(),
        child: GoogleMap(
            mapType: MapType.normal,
            initialCameraPosition: kSantoDomingo,
            onMapCreated: (gController) {
              stream.forEach((value) => newLocationUpdate(value));
              controller.complete(gController);
              //Complete the future GoogleMapController
            }),
      ),
    );
  }

  void newLocationUpdate(LatLng latLng) {
    var marker = RippleMarker(
        markerId: kMarkerId,
        position: latLng,
        ripple: true,
        onTap: () {
          print('Tapped! $latLng');
        });
    setState(() => markers[kMarkerId] = marker);
  }
}
